### PR TITLE
Add Db2Monitoring to verify that all packages in the database have been monitored

### PR DESCRIPTION
**Pull request by scottbommarito**
https://github.com/NuGet/Engineering/issues/318

This is an additional job for the V3 monitoring pipeline that will verify that all packages changes have been processed and saved by `MonitoringProcessorJob`.

The job works as follows:
- load a cursor on the database `LastEdited` and audit entries
    - these cursors are stored in the `v3-monitoring` container (`gallerycursor.json` and `deletedcursor.json`)
- fetch the next `Top` packages from the database and the next `Top` audit entries
- for each package, check the monitoring blobs
    - if the monitoring blob is less recent, queue a message to run validations on the package
- update the database and audit entry cursors
- if both the database and audit entry cursors are up to date, reset to the beginning of time and reprocess all packages

By scanning through the database and deleted audit records and then comparing these to the monitoring blobs, we will now be able to determine automatically whether or not
- all package changes have been flown to V3
- `MonitoringProcessorJob` has not skipped any package changes that did flow to V3

Locally, I was able to process 550 packages in roughly 15 seconds, which means that we should be able to process the entirety of PROD every 12-13 hours.

### CODE REVIEW

File: src/Catalog/Helpers/Db2CatalogProjection.cs [hunk 1]
```diff
@@ -34,6 +34,7 @@ public FeedPackageDetails ReadFeedPackageDetailsFromDataReader(DbDataReader data
 
             var packageId = dataReader[Db2CatalogProjectionColumnNames.PackageId].ToString();
             var normalizedPackageVersion = dataReader[Db2CatalogProjectionColumnNames.NormalizedVersion].ToString();
+            var fullPackageVersion = dataReader[Db2CatalogProjectionColumnNames.FullVersion].ToString();
```
Comment (scottbommarito) [inline src/Catalog/Helpers/Db2CatalogProjection.cs]: Unfortunately, monitoring blobs are stored based on a package's "full version" and not their "normalized version". This means that I needed to update the package fetching code used by `Db2Catalog` to grab this information as well. I felt like there isn't very much risk in this but I understand there may be some concerns.

File: src/NuGet.Services.Metadata.Catalog.Monitoring/Model/PackageTimestampMetadata.cs [hunk 2]
```diff
@@ -53,11 +53,33 @@ public static PackageTimestampMetadata CreateForMissingPackage(DateTime? deleted
             };
         }
 
+        /// <remarks>
+        /// In the past, the catalog entries referred to the CDN's host instead our DNS.
+        /// If we encounter one of these outdated hosts, we should hit the proper host instead.
+        /// </remarks>
+        private static readonly IDictionary<string, string> _catalogEntryUriHostMap =
```
Comment (scottbommarito) [inline src/NuGet.Services.Metadata.Catalog.Monitoring/Model/PackageTimestampMetadata.cs]: Temporary mitigation for https://github.com/NuGet/Engineering/issues/2505

**Comment by scottbommarito** [comment]
I'll be doing a design review for this and the other recent monitoring changes tomorrow.

File: src/NuGet.Services.Metadata.Catalog.Monitoring/Monitoring/PackageStatusOutdatedCheckSource.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace NuGet.Services.Metadata.Catalog.Monitoring.Monitoring
+{
+    public interface IPackageStatusOutdatedCheckSource
```
Comment (agr) [inline src/NuGet.Services.Metadata.Catalog.Monitoring/Monitoring/PackageStatusOutdatedCheckSource.cs]: Documentation is missing for new classes/interfaces and methods.

Comment (scottbommarito) [reply src/NuGet.Services.Metadata.Catalog.Monitoring/Monitoring/PackageStatusOutdatedCheckSource.cs]: done!

File: src/Catalog/Helpers/FeedPackageDetails.cs [hunk 4]
```diff
@@ -25,14 +26,16 @@ public sealed class FeedPackageDetails
             DateTime lastEditedDate,
             DateTime publishedDate,
             string packageId,
-            string packageVersion)
+            string packageVersion,
```
Comment (loic-sharma) [inline src/Catalog/Helpers/FeedPackageDetails.cs]: Maybe we should rename this to `normalizedVersion`? Same comment to line 15

File: src/NuGet.Services.Metadata.Catalog.Monitoring/Monitoring/PackageStatusOutdatedCheckSource.cs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        public async Task<IReadOnlyCollection<PackageStatusOutdatedCheck>> GetPackagesToCheckAsync(int top, CancellationToken cancellationToken)
+        {
+            await _cursor.LoadAsync(cancellationToken);
+            var packages = await GetPackagesToCheckAsync(_cursor.Value, top, cancellationToken);
+            _pendingCursorValue = packages.Any() 
+                ? packages.Max(p => GetCursorValue(p)) 
+                : (DateTime?)null;
+
+            return packages
+                .Select(p => GetPackageStatusOutdatedCheck(p))
```
Comment (loic-sharma) [inline src/NuGet.Services.Metadata.Catalog.Monitoring/Monitoring/PackageStatusOutdatedCheckSource.cs]:
```suggestion
                .Select(GetPackageStatusOutdatedCheck)
```

File: src/NuGet.Services.Metadata.Catalog.Monitoring/Monitoring/PackageStatusOutdatedCheckSource.cs [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        {
+            _cursor = cursor ?? throw new ArgumentNullException(nameof(cursor));
+        }
+
+        public async Task<IReadOnlyCollection<PackageStatusOutdatedCheck>> GetPackagesToCheckAsync(int top, CancellationToken cancellationToken)
+        {
+            await _cursor.LoadAsync(cancellationToken);
+            var packages = await GetPackagesToCheckAsync(_cursor.Value, top, cancellationToken);
+            _pendingCursorValue = packages.Any() 
+                ? packages.Max(p => GetCursorValue(p))
```
Comment (loic-sharma) [inline src/NuGet.Services.Metadata.Catalog.Monitoring/Monitoring/PackageStatusOutdatedCheckSource.cs]:
```suggestion
                ? packages.Max(GetCursorValue) 
```

File: src/NuGet.Services.Metadata.Catalog.Monitoring/Monitoring/PackageStatusOutdatedCheckSource.cs [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+namespace NuGet.Services.Metadata.Catalog.Monitoring.Monitoring
+{
+    /// <summary>
+    /// Fetches <see cref="PackageStatusOutdatedCheck"/>s from a source and maintains a cursor on it.
+    /// </summary>
+    public interface IPackageStatusOutdatedCheckSource
+    {
+        Task<IReadOnlyCollection<PackageStatusOutdatedCheck>> GetPackagesToCheckAsync(int top, CancellationToken cancellationToken);
+        Task MarkPackagesCheckedAsync(CancellationToken cancellationToken);
+        Task ResetAsync(CancellationToken cancellationToken);
```
Comment (loic-sharma) [inline src/NuGet.Services.Metadata.Catalog.Monitoring/Monitoring/PackageStatusOutdatedCheckSource.cs]: Don't forget the comments on methods 😄

Comment (scottbommarito) [reply src/NuGet.Services.Metadata.Catalog.Monitoring/Monitoring/PackageStatusOutdatedCheckSource.cs]: now they are added haha

File: src/Catalog/Helpers/FeedPackageDetails.cs (hunk 4 above)
Comment (scottbommarito) [reply src/Catalog/Helpers/FeedPackageDetails.cs]: I worry that doing this rename will lead to a number of changes all over the project. I will check it out and see how much is impacted. If it generates too many changes, I think adding an XML doc that explains that it's the normalized version would be sufficient.

File: src/Catalog/Helpers/Db2CatalogProjection.cs (hunk 1 above)
Comment (xavierdecoster) [reply src/Catalog/Helpers/Db2CatalogProjection.cs]: This is fine

File: src/NuGet.Services.Metadata.Catalog.Monitoring/Monitoring/DatabasePackageStatusOutdatedCheckSource.cs [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    /// Fetches <see cref="PackageStatusOutdatedCheck"/> from <see cref="IGalleryDatabaseQueryService.GetPackagesEditedSince(DateTime, int)"/>.
+    /// The <see cref="PackageStatusOutdatedCheck"/>s fetched represent the latest state for existing packages.
+    /// </summary>
+    public class DatabasePackageStatusOutdatedCheckSource : PackageStatusOutdatedCheckSource<FeedPackageDetails>
+    {
+        private readonly IGalleryDatabaseQueryService _galleryDatabaseQueryService;
+
+        public DatabasePackageStatusOutdatedCheckSource(
+            ReadWriteCursor cursor,
+            IGalleryDatabaseQueryService galleryDatabaseQueryService)
```
Comment (xavierdecoster) [inline src/NuGet.Services.Metadata.Catalog.Monitoring/Monitoring/DatabasePackageStatusOutdatedCheckSource.cs]: Should we call the argument `galleryDatabase` for consistency with other places where we inject `IGalleryDatabaseQueryService`?

File: tests/CatalogTests/Helpers/CatalogWriterHelperTests.cs [hunk 9]
```diff
@@ -405,7 +405,8 @@ internal WritePackageDetailsToCatalogAsyncTest()
                     UtcNow.AddMinutes(-30),
                     UtcNow.AddMinutes(-15),
                     packageId: "a",
-                    packageVersion: "1.0.0");
+                    packageVersion: "1.0.0",
+                    packageFullVersion: "1.0.0");
```
Comment (xavierdecoster) [inline tests/CatalogTests/Helpers/CatalogWriterHelperTests.cs]: Perhaps make sure `packageVersion` and `packageFulVersion` differ?
e.g. by making `packageFullVersion` a SemVer2 version with metadata.

Comment (xavierdecoster) [reply tests/CatalogTests/Helpers/CatalogWriterHelperTests.cs]: Same for `Db2CatalogProjectionTests` where you added `packageFullVersion`.

File: src/NuGet.Services.Metadata.Catalog.Monitoring/Monitoring/DatabasePackageStatusOutdatedCheckSource.cs (hunk 8 above)
Comment (scottbommarito) [reply src/NuGet.Services.Metadata.Catalog.Monitoring/Monitoring/DatabasePackageStatusOutdatedCheckSource.cs]: Yup, my bad--keep forgetting what we call it!.

File: src/Catalog/Helpers/FeedPackageDetails.cs (hunk 4 above)
Comment (scottbommarito) [reply src/Catalog/Helpers/FeedPackageDetails.cs]: Didn't effect nearly as much as I expected. Will do.

**Comment by scottbommarito** [comment]
Will be merging this tomorrow after I test it on DEV!

**Pull request merged**